image/ani.cpp

```diff
@@ -0,0 +1,304 @@
+		warning("AniDecoder::parseTopLevelChunk: Top-level chunk isn't RIFF");
+		return false;
+	}
+
+	return parseRIFFContainer(stream, chunk, Common::Functor2Mem<const RIFFContainerDef &, Common::SeekableReadStream &, bool, AniDecoder>(this, &AniDecoder::parseTopLevelContainer));
+}
+
+bool AniDecoder::parseTopLevelContainer(const RIFFContainerDef &container, Common::SeekableReadStream &stream) {
+	if (container.id == MKTAG('A', 'C', 'O', 'N'))
+		return parseRIFFChunks(stream, Common::Functor2Mem<const RIFFChunkDef &, Common::SeekableReadStream &, bool, AniDecoder>(this, &AniDecoder::parseSecondLevelChunk));
+
+	warning("AniDecoder::parseTopLevelContainer: Top-level container isn't ACON");
+	return false;
+}
+
+bool AniDecoder::parseSecondLevelChunk(const RIFFChunkDef &chunk, Common::SeekableReadStream &stream) {
+	if (chunk.id == MKTAG('L', 'I', 'S', 'T'))
+		return parseRIFFContainer(stream, chunk, Common::Functor2Mem<const RIFFContainerDef &, Common::SeekableReadStream &, bool, AniDecoder>(this, &AniDecoder::parseListContainer));
+
+	if (chunk.id == MKTAG('a', 'n', 'i', 'h'))
+		return parseAnimHeaderChunk(chunk, stream);
+
+	if (chunk.id == MKTAG('s', 'e', 'q', ' '))
+		return parseSeqChunk(chunk, stream);
+
+	if (chunk.id == MKTAG('r', 'a', 't', 'e'))
+		return parseRateChunk(chunk, stream);
+
+	return true;
+}
+
+bool AniDecoder::parseListContainer(const RIFFContainerDef &container, Common::SeekableReadStream &stream) {
+	if (container.id == MKTAG('f', 'r', 'a', 'm'))
+		return parseRIFFChunks(stream, Common::Functor2Mem<const RIFFChunkDef &, Common::SeekableReadStream &, bool, AniDecoder>(this, &AniDecoder::parseIconChunk));
+
+	return true;
+}
+
+bool AniDecoder::parseAnimHeaderChunk(const RIFFChunkDef &chunk, Common::SeekableReadStream &stream) {
+	byte animHeader[36];
+	for (byte &b : animHeader)
+		b = 0;
+
+	uint32 amountToRead = 36;
+	if (chunk.size < amountToRead)
+		amountToRead = chunk.size;
+
+	if (amountToRead > 0 && stream.read(animHeader, amountToRead) != amountToRead) {
+		warning("AniDecoder::parseAnimHeaderChunk: Read failed");
+		return false;
+	}
+
+	uint32 structSize = READ_LE_UINT32(animHeader);
+	if (structSize < 36) {
+		for (uint i = structSize; i < 36; i++)
+			animHeader[i] = 0;
+	}
+
+	_metadata.numFrames = READ_LE_UINT32(animHeader + 4);
+	_metadata.numSteps = READ_LE_UINT32(animHeader + 8);
+	_metadata.width = READ_LE_UINT32(animHeader + 12);
+	_metadata.height = READ_LE_UINT32(animHeader + 16);
+	_metadata.bitCount = READ_LE_UINT32(animHeader + 20);
+	_metadata.numPlanes = READ_LE_UINT32(animHeader + 24);
+	_metadata.perFrameDelay = READ_LE_UINT32(animHeader + 28);
+
+	uint32 flags = READ_LE_UINT32(animHeader + 32);
+	_metadata.isCURFormat = ((flags & 1) != 0);
+	_metadata.haveSeqData = ((flags & 2) != 0);
+
+	return true;
+}
+
+bool AniDecoder::parseSeqChunk(const RIFFChunkDef &chunk, Common::SeekableReadStream &stream) {
+	uint32 numFrames = chunk.size / 4u;
+
+	if (numFrames > 1000u) {
```

Comment:
Why this magic number?

Reply:
I guess I could increase it but it's supposed to just be a safeguard against reading a huge frame count that likely indicates some kind of internal error or corruption.